### PR TITLE
mgr/cephadm: store agent metadata in its own cache

### PR DESCRIPTION
**Pull request by adk3798**
To avoid having the host cache getting too big
due to having to store this additional info

Fixes: https://tracker.ceph.com/issues/53624

Signed-off-by: Adam King <adking@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/migrations.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 if TYPE_CHECKING:
     from .module import CephadmOrchestrator
 
-LAST_MIGRATION = 5
+LAST_MIGRATION = 6
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/migrations.py:15]: Thinking. There was no production cluster with this the agent enabled. IMO we can avoid a migration here. We don't support migrations between development version.

Comment (adk3798) [reply src/pybind/mgr/cephadm/migrations.py:15]: alright, I'll take that bit out

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
```
714]: Failed while placing agent.smithi031 on smithi031: cephadm exited with an error code: 1, stderr: This is a development version of cephadm.
      For information regarding the latest stable release:
          https://docs.ceph.com/docs/pacific/cephadm/install
      Deploy daemon agent.smithi031 ...
      Non-zero exit code 1 from systemctl enable --now ceph-b041d42a-7a06-11ec-8c35-001a4aab830c@agent.smithi031.service
      systemctl: stderr Job for ceph-b041d42a-7a06-11ec-8c35-001a4aab830c@agent.smithi031.service failed because the control process exited with error code.
      systemctl: stderr See "systemctl status ceph-b041d42a-7a06-11ec-8c35-001a4aab830c@agent.smithi031.service" and "journalctl -xe" for details.
      Traceback (most recent call last):
        File "/var/lib/ceph/b041d42a-7a06-11ec-8c35-001a4aab830c/cephadm.5974b2b6c251143d78f026a5400c5b6618bd61dede4903945a2d4a8f8694b0ac", line 8612, in <module>
          main()
        File "/var/lib/ceph/b041d42a-7a06-11ec-8c35-001a4aab830c/cephadm.5974b2b6c251143d78f026a5400c5b6618bd61dede4903945a2d4a8f8694b0ac", line 8600, in main
          r = ctx.func(ctx)
        File "/var/lib/ceph/b041d42a-7a06-11ec-8c35-001a4aab830c/cephadm.5974b2b6c251143d78f026a5400c5b6618bd61dede4903945a2d4a8f8694b0ac", line 1938, in _default_image
          return func(ctx)
        File "/var/lib/ceph/b041d42a-7a06-11ec-8c35-001a4aab830c/cephadm.5974b2b6c251143d78f026a5400c5b6618bd61dede4903945a2d4a8f8694b0ac", line 5367, in command_deploy
          uid, gid, ports=daemon_ports)
        File "/var/lib/ceph/b041d42a-7a06-11ec-8c35-001a4aab830c/cephadm.5974b2b6c251143d78f026a5400c5b6618bd61dede4903945a2d4a8f8694b0ac", line 2894, in deploy_daemon
          cephadm_agent.deploy_daemon_unit(config_js)
        File "/var/lib/ceph/b041d42a-7a06-11ec-8c35-001a4aab830c/cephadm.5974b2b6c251143d78f026a5400c5b6618bd61dede4903945a2d4a8f8694b0ac", line 3814, in deploy_daemon_unit
          call_throws(self.ctx, ['systemctl', 'enable', '--now', self.unit_name()])
        File "/var/lib/ceph/b041d42a-7a06-11ec-8c35-001a4aab830c/cephadm.5974b2b6c251143d78f026a5400c5b6618bd61dede4903945a2d4a8f8694b0ac", line 1615, in call_throws
          raise RuntimeError(f'Failed command: {" ".join(command)}: {s}')
      RuntimeError: Failed command: systemctl enable --now ceph-b041d42a-7a06-11ec-8c35-001a4aab830c@agent.smithi031.service: Job for ceph-b041d42a-7a06-11ec-8c35-001a4aab830c@agent.smithi031.service failed bec>
      See "systemctl status ceph-b041d42a-7a06-11ec-8c35-001a4aab830c@agent.smithi031.service" and "journalctl -xe" for details.
```
sadly we don't have the agent logs. http://qa-proxy.ceph.com/teuthology/swagner-2022-01-20_15:11:07-orch:cephadm-wip-swagner-testing-2022-01-20-1235-distro-default-smithi/6629022/remote/smithi031/log/b041d42a-7a06-11ec-8c35-001a4aab830c/ . An idea would be to 

https://github.com/ceph/ceph/blob/0c2dfd9bd38d59b5fd994c1b6ef1f0bef45352aa/qa/tasks/cephadm.py#L265

that fetches the agent logs.

**Comment by adk3798** [comment]
```
Jan 24 22:58:07 smithi049 systemd[1]: Starting cephadm agent for cluster c8e864c0-7d68-11ec-8c35-001a4aab830c...
Jan 24 22:58:07 smithi049 bash[37116]: /bin/bash: /var/lib/ceph/c8e864c0-7d68-11ec-8c35-001a4aab830c/agent.smithi049/unit.run: No such file or directory
Jan 24 22:58:07 smithi049 systemd[1]: ceph-c8e864c0-7d68-11ec-8c35-001a4aab830c@agent.smithi049.service: Control process exited, code=exited status=127
Jan 24 22:58:07 smithi049 systemd[1]: ceph-c8e864c0-7d68-11ec-8c35-001a4aab830c@agent.smithi049.service: Failed with result 'exit-code'.
Jan 24 22:58:07 smithi049 systemd[1]: Failed to start cephadm agent for cluster c8e864c0-7d68-11ec-8c35-001a4aab830c.
```
Teuthology failure ended up being an issue with writing the unit run file. Should be good now.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-01-26_05:20:46-orch:cephadm-wip-adk-testing-2022-01-25-2134-distro-basic-smithi/

6 dead ingress jobs tracked by https://tracker.ceph.com/issues/53807
1 failed job related to deploying cephfs-mirror tracked by https://tracker.ceph.com/issues/54029

**Pull request merged**